### PR TITLE
mm rewrite

### PR DESCRIPTION
**Pull request by V01D-NULL**
Finished the primary base allocator.
It consists of 3 functions
- `init` ~ Just pass in the the mmap tag given to us by limine and we're good to go.
- `allocate`
- `deallocate`

`allocate`'s parameter is the number of pages it should allocate, it returns a linked list in case the allocated regions are non contiguous. Upon allocation the first element in the linked list is __RESERVED__ and not intended for use by anything outside of the allocator.

Allocation and deallocation demo:
```c++
// unpack() checks if the head node is valid
// returns NULL if it's not, otherwise it
// returns the *real* head of the list (It's the reserved element I mentioned above)
auto ptr = allocate(2)->unpack(); // Allocate 2 pages- Panics on failure. 

if  (ptr == nullptr)
     panic("The head node was not identified (this is weird and shouldn't really happen but you never know)");

// You are required to make a copy
// of the linked list if you wish to 
// deallocate it *and* iterate through it's contents.
// (`ptr = ptr->next ` would mess up the entire
// structure and loose information when passing `ptr` to deallocate)
auto ptr2 = ptr;

// Iterate through all pages allocated
// (alternatively you can iterate through less)
while (ptr != nullptr)
{
    printf("Page allocated at %X\n", (size_t)ptr->addr); //ptr->addr is a void*
    ptr = ptr->next;
}

deallocate(ptr2); // Note the usage of ptr*2*, not ptr.
//Alternatively one could reverse the changes performed to ptr
// but I'll just demonstrate the copy method for the sake
// of simplicity. Large allocations should *definitely* avoid the copy method.
```

### CODE REVIEW

File: arch/x86_64/kernel/fb/double-buffering.cpp [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+void DoubleBuffering::init_buffers() {
+    //Todo: Alloc buffers once we have the mm bug fixed
+    //alloc(this->front)
+    //alloc(this->back);
+
+    front.is_front = true;
+    back.is_front = false;
+
+#if DEBUG
```
Comment (Yuhanun) [inline arch/x86_64/kernel/fb/double-buffering.cpp:18]: Isn't it NDEBUG?

File: arch/x86_64/kernel/fb/double-buffering.cpp [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    front.id = "front";
+    back.id = "back";
+#endif
+}
+
+/* Render the contents of a buffer to the screen. */
+void DoubleBuffering::render_buffer(tty_render_buffer& buffer) {
+    libkern::Vec2 dimensions = firefly::drivers::vbe::fb_dimensions();
+    memcpy(
```
Comment (Yuhanun) [inline arch/x86_64/kernel/fb/double-buffering.cpp:27]: std::copy?

File: arch/x86_64/kernel/fb/double-buffering.cpp [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+void DoubleBuffering::buffer_pixel(size_t x, size_t y, uint32_t color) {
+    get_active().buffer[(y * (drivers::vbe::fb_pitch() / sizeof(uint32_t))) + x] = color;
+}
+
+/* Swap both buffers and render the previous back buffer to the screen */
+void DoubleBuffering::swap_buffers() {
+    front.is_front = !front.is_front;
+    back.is_front = !back.is_front;
+    back.is_front ? render_buffer(front) : render_buffer(back);
```
Comment (Yuhanun) [inline arch/x86_64/kernel/fb/double-buffering.cpp:42]: render_buffer(back.is_front ? front : back);

File: arch/x86_64/kernel/fb/double-buffering.cpp [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+}
+
+/* Return the active back buffer to draw to it */
+tty_render_buffer& DoubleBuffering::get_active() {
+    if (front.is_front)
```
Comment (Yuhanun) [inline arch/x86_64/kernel/fb/double-buffering.cpp:47]: return front.is_front ? back : front;

Comment (Yuhanun) [reply arch/x86_64/kernel/fb/double-buffering.cpp:47]: to be fair we should do all these branchless but it is what it is

File: arch/x86_64/kernel/memory-manager/primary/primary_phys.cpp [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+
+#include <stl/cstdlib/stdio.h>
+
+#include "x86_64/libk++/align.h"
+#include "x86_64/trace/strace.hpp"
+
+namespace firefly::kernel::mm::primary {
+static libkern::Bitmap bitmap;
+static uint32_t *arena;
+static int64_t linked_list_allocation_base;  //Base address for the linked list structure (Should never be freed, may be reused)
```
Comment (Yuhanun) [inline arch/x86_64/kernel/memory-manager/primary/primary_phys.cpp:11]: This should just be a list tho, not a linked list there's no point. but yea

Comment (Yuhanun) [reply arch/x86_64/kernel/memory-manager/primary/primary_phys.cpp:11]: like you have 4096 bytes, why not just say "every 8 bytes is an address" and then return it by casting it to a void*[] and a size_t

File: arch/x86_64/libk++/bitmap.cpp [hunk 6]
```diff
@@ -0,0 +1,93 @@
+#include "x86_64/libk++/bitmap.h"
+#include <stl/cstdlib/stdio.h>
+#include <stl/cstdlib/cstring.h>
+
+namespace firefly::libkern
+{
+    void Bitmap::init(uint32_t *bitmap, size_t size)
```
Comment (Yuhanun) [inline arch/x86_64/libk++/bitmap.cpp:7]: actual C++ <3

**Review by Yuhanun — APPROVED** [review]
if you can, change the returned linked list to an array, if not then it's okay <3 we will optimize later!

File: arch/x86_64/kernel/fb/double-buffering.cpp (hunk 4 above)
Comment (V01D-NULL) [reply arch/x86_64/kernel/fb/double-buffering.cpp:47]: Ignore everything related to double buffering, I wasn't smart about it then but I have a better way of implementing this now. It'll be branchless btw

File: arch/x86_64/kernel/fb/double-buffering.cpp (hunk 1 above)
Comment (V01D-NULL) [reply arch/x86_64/kernel/fb/double-buffering.cpp:18]: NDEBUG?
You'd just add -D DEBUG to the compiler flags to enable that printf code

File: arch/x86_64/libk++/bitmap.cpp (hunk 6 above)
Comment (V01D-NULL) [reply arch/x86_64/libk++/bitmap.cpp:7]: Was that Sarcasm / did you want me to write a constructor for it?

**Pull request merged**